cpp/src/arrow/flight/sql/driver.cc

```diff
@@ -0,0 +1,1877 @@
+      }
+
+      if (depth == ADBC_OBJECT_DEPTH_ALL || depth >= ADBC_OBJECT_DEPTH_DB_SCHEMAS) {
+        ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error, catalog_db_schemas_builder.Append());
+      } else {
+        ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                 catalog_db_schemas_builder.AppendNull());
+        continue;
+      }
+
+      auto it = table_index.find(cur_catalog_name);
+      if (it == table_index.end()) continue;
+
+      for (const auto& schema_item : it->second) {
+        ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                 catalog_db_schemas_items_builder->Append());
+        const std::optional<std::string>& cur_schema_name = schema_item.first;
+
+        if (cur_schema_name) {
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                   db_schema_name_builder->Append(*cur_schema_name));
+        } else {
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error, db_schema_name_builder->AppendNull());
+        }
+
+        if (depth == ADBC_OBJECT_DEPTH_ALL || depth >= ADBC_OBJECT_DEPTH_TABLES) {
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error, db_schema_tables_builder->Append());
+        } else {
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                   db_schema_tables_builder->AppendNull());
+          continue;
+        }
+
+        std::pair<int64_t, int64_t> schema_tables_range = schema_item.second;
+        for (int64_t i = schema_tables_range.first; i < schema_tables_range.second; i++) {
+          const std::string_view table_name = GetNotNull<StringType>(*table_data, 2, i);
+          const std::string_view table_type = GetNotNull<StringType>(*table_data, 3, i);
+          const std::string_view table_schema = GetNotNull<BinaryType>(*table_data, 4, i);
+
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                   db_schema_tables_items_builder->Append());
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                   table_name_builder->Append(table_name));
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                   table_type_builder->Append(table_type));
+
+          if (depth == ADBC_OBJECT_DEPTH_COLUMNS) {
+            ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error, table_columns_builder->Append());
+
+            std::shared_ptr<Schema> schema;
+            io::BufferReader reader((Buffer(table_schema)));
+            ipc::DictionaryMemo memo;
+            ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                     ipc::ReadSchema(&reader, &memo).Value(&schema));
+
+            int32_t ordinal_position = 1;
+            for (const std::shared_ptr<Field>& field : schema->fields()) {
+#ifdef ARROW_COMPUTE
+              if (column_name && std::strlen(column_name) > 0) {
+                Datum result;
+                arrow::compute::MatchSubstringOptions options(column_name);
```

Comment:
It's gonna be very slow to initialize a kernel for each column name in each database table... once we extract the required facilities (see other comment) we'll be able to make this faster, probably.